File: modules/biscale.c
#include <stdlib.h>
#include "utone.h"

int ut_biscale_create(ut_biscale **p)
{
    *p = malloc(sizeof(ut_biscale));
    return UT_OK;
}

int ut_biscale_destroy(ut_biscale **p)
{
    free(*p);
    return UT_OK;
}

int ut_biscale_init(ut_data *ut, ut_biscale *p)
{
    p->min = 0;
    p->max = 1;
    return UT_OK;
}

int ut_biscale_compute(ut_data *ut, ut_biscale *p, UTFLOAT *in, UTFLOAT *out)
{
    *out = p->min + (*in + 1.0) / 2.0 * (p->max - p->min);
    return UT_OK;
}

File: modules/data/biscale.lua
sptbl["biscale"] = {

    files = {
        module = "biscale.c",
        header = "biscale.h",
        example = "ex_biscale.c",
    },

    func = {
        create = "ut_biscale_create",
        destroy = "ut_biscale_destroy",
        init = "ut_biscale_init",
        compute = "ut_biscale_compute",
    },

    params = {
        optional = {
            {
                name = "min",
                type = "UTFLOAT",
                description = "Minimum value to scale to.",
                default = 0
            },
            {
                name = "max",
                type = "UTFLOAT",
                description ="Maximum value to scale to.",
                default = 1
            },
        }
    },

    modtype = "module",

    description = [[Bipolar Scale

    This module scales from bipolar [-1, 1] to another range defined by min and max.
]],

    ninputs = 1,
    noutputs = 1,

    inputs = {
        {
            name = "in",
            description = "Signal to be scaled."
        },
    },

    outputs = {
        {
            name = "out",
            description = "Scaled signal out"
        },
    }

}
